analysis/src/SignatureHelp.ml

```diff
@@ -0,0 +1,213 @@
+                |> Option.get
+              in
+              let endOffset =
+                argumentTypeExpr.ptyp_loc |> Loc.end_
+                |> CompletionFrontEnd.positionToOffset label
+                |> Option.get
+              in
+              (* The AST locations does not account for "=?" of optional arguments, so add that to the offset here if needed. *)
+              let endOffset =
+                match argumentLabel with
+                | Asttypes.Optional _ -> endOffset + 2
+                | _ -> endOffset
+              in
+              extractParams nextFunctionExpr
+                (params @ [(startOffset, endOffset)])
+            | _ -> params
+          in
+          extractParams expr []
+        | _ -> []
+      in
+
+      if debug then
+        Printf.printf "extracted params: \n%s\n"
+          (parameters
+          |> List.map (fun (start, end_) ->
+                 String.sub label start (end_ - start))
+          |> SharedTypes.list);
+
+      (* Figure out the active parameter *)
+      let activeParameter =
+        match argAtCursor with
+        | None -> (
+          (* If a function only has one, unlabelled argument, we can safely assume that's active whenever we're in the signature help for that function,
+             even if we technically didn't find anything at the cursor (which we don't for empty expressions). *)
+          match args with
+          | [(Nolabel, _)] -> Some 0
+          | _ -> None)
+        | Some (Unlabelled unlabelledArgumentIndex) ->
+          let index = ref 0 in
+          args
+          |> List.find_map (fun (label, _) ->
+                 match label with
+                 | Asttypes.Nolabel when !index = unlabelledArgumentIndex ->
+                   Some !index
+                 | _ ->
+                   index := !index + 1;
+                   None)
+        | Some (Labelled name) ->
+          let index = ref 0 in
+          args
+          |> List.find_map (fun (label, _) ->
+                 match label with
+                 | (Asttypes.Labelled labelName | Optional labelName)
+                   when labelName = name ->
+                   Some !index
+                 | _ ->
+                   index := !index + 1;
+                   None)
+      in
+      Some
+        {
+          Protocol.signatures =
+            [
+              {
+                label;
+                parameters =
+                  parameters
+                  |> List.map (fun params -> {Protocol.label = params});
+                documentation =
+                  (match List.nth_opt docstring 0 with
+                  | None -> None
+                  | Some docs -> Some {Protocol.kind = "markdown"; value = docs});
+              };
+            ];
+          activeSignature = Some 0;
+          activeParameter =
+            (match activeParameter with
+            | None -> Some (-1)
+            | activeParameter -> activeParameter);
+        }
```

Comment:
Not going into too many comments on this function, rather some high level observations:

1 try to split up into smaller piece and make the toplevel logic short and clear

2 check if enough has been reused of existing logic, and avoid repetitions here -- not clear that there are repetitions, not clear either that there are not

Reply:
Got it! Will revisit.